third_party/move/move-vm/types/src/values/values_impl.rs

```diff
@@ -3383,12 +3408,28 @@ impl<'d, C: CustomDeserializer> serde::de::DeserializeSeed<'d>
         self,
         deserializer: D,
     ) -> Result<Self::Value, D::Error> {
-        let field_layouts = self.layout.fields();
-        let fields = deserializer.deserialize_tuple(
-            field_layouts.len(),
-            StructFieldVisitor(self.custom_deserializer, field_layouts),
-        )?;
-        Ok(Struct::pack(fields))
+        match &self.layout {
+            MoveStructLayout::Runtime(field_layouts) => {
+                let fields = deserializer.deserialize_tuple(
+                    field_layouts.len(),
+                    StructFieldVisitor(self.custom_deserializer, field_layouts),
+                )?;
+                Ok(Struct::pack(fields))
+            },
+            MoveStructLayout::RuntimeVariants(variant_layouts) => {
+                let (tag, mut fields) = deserializer.deserialize_seq(StructVariantVisitor(
+                    self.custom_deserializer,
+                    variant_layouts,
+                ))?;
+                fields.insert(0, Value::u16(tag));
```

Comment:
Is it worth keeping tag separate? Why not `StructVariantVisitor` produces tag as the first element directly? This insertion at 0 can be expensive.